lib/caputils.c

```diff
@@ -14,32 +14,40 @@
  * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
  */
 
-#include <stdio.h>
+#include <sys/prctl.h>
+#include <limits.h>
 
 #include "caputils.h"
 #include "pathnames.h"
 
-int cap_last_cap(void)
+static int test_cap(int cap)
 {
-	/* CAP_LAST_CAP is untrustworthy. */
-	static int ret = -1;
-	int matched;
-	FILE *f;
-
-	if (ret != -1)
-		return ret;
-
-	f = fopen(_PATH_PROC_CAPLASTCAP, "r");
-	if (!f) {
-		ret = CAP_LAST_CAP;	/* guess */
-		return ret;
-	}
+	/* prctl returns 0 or 1 for valid caps, -1 otherwise */
+	return prctl(PR_CAPBSET_READ, cap, 0, 0, 0) >= 0;
+}
 
-	matched = fscanf(f, "%d", &ret);
-	fclose(f);
+/* this cache makes cap_last_cap MT-unsafe */
+static int last_cap = 0;
 
-	if (matched != 1)
-		ret = CAP_LAST_CAP;	/* guess */
+int cap_last_cap(void)
+{
+	/* cached last_cap value */
+	if (last_cap) return last_cap;
+
+	/* binary search over capabilities */
+	int cap0 = 0, cap1 = INT_MAX;
+	while (1) {
```

Comment:
`while (cap0 + 1 < cap1)`, make the criteria for termination as loose as possible.

Reply:
On second thought, initialize `cap = INT_MAX` outside the loop and move the test to
```
while (cap0 < cap)
```
This will take care of the degenerate average because, for `cap0 = N` and `cap0 <= N + 1` for any `N`, the average `cap <= cap0`.

Reply:
> `cap = INT_MAX`

this is adding another iteration for most cases, where `cap1` is reassigned `INT_MAX`, to help with the case where `last_cap == INT_MAX`, isn't it? I think the intent is clearer with `cap = INT_MAX / 2`...

Reply:
I see that starting with `cap = INT_MAX` means I always know that `cap1` is invalid, which allows me to use `last_cap = cap` in the end.

Reply:
It doesn't add another iteration, because in the first iteration you are doing `cap = (cap0 + cap1) / 2` before anything else, so you never actually test `INT_MAX` in the loop.